docs/proposals/08-shoot-apiserver-via-sni.md

```diff
@@ -0,0 +1,396 @@
+- https://github.com/xtaci/kcptun (transport is UDP-based)
+- OpenVPN is discarded as the performance is abysmal
+
+`chisel` is the fastest and offers the least latency when testing under different conditions. The architecture is simple:
+
+![overview](https://docs.google.com/drawings/d/1p53VWxzGNfy8rjr-mW8pvisJmhkoLl82vAgctO_6f1w/pub?w=960&h=720)
+
+For the Gardener implementation, the `clisel-client` will run as a `DaemonSet` on the `Shoot` cluster and a dedicated `chisel-server` is provisioned for each `Shoot` control plane. To reduce the amount of records we create, it listens on a dedicated port `9443`.
+
+An updated version of the previous SNI configuration:
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  name: kube-apiserver-gateway
+  namespace: shoot-namespace
+spec:
+  selector:
+    istio: ingressgateway
+  servers:
+  - port:
+      number: 443
+      name: tls
+      protocol: TLS
+    tls:
+      mode: PASSTHROUGH
+    hosts:
+    - api.<shoot>.<seed>.<domain>
+    - api.internal.<shoot>.<seed>.<domain>
+  - port:
+      number: 9443
+      name: tls
+      protocol: TLS
+    tls:
+      mode: PASSTHROUGH
+    hosts:
+    - api.internal.<shoot>.<seed>.<domain>
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: kube-apiserver
+  namespace: shoot-namespace
+spec:
+  hosts:
+  - api.<shoot>.<seed>.<domain>
+  - api.internal.<shoot>.<seed>.<domain>
+  gateways:
+  - kube-apiserver-gateway
+  tls:
+  - match:
+    - port: 443
+      sniHosts:
+      - api.<shoot>.<seed>.<domain>
+      - api.internal.<shoot>.<seed>.<domain>
+    route:
+    - destination:
+        host: kube-apiserver.shoot-namespace.svc.cluster.local
+        port:
+          number: 443
+  - match:
+    - port: 9443
+      sniHosts:
+      - api.internal.<shoot>.<seed>.<domain>
+    route:
+    - destination:
+        host: proxy.shoot-namespace.svc.cluster.local
+        port:
+          number: 443
+```
+
+Chisel works by creating logical TCP connections via SSH, but it uses only one TCP connection per client/Shoot Node.
+With this approach a Chisel server (with appropriate replica count) can be deployed and it'll serve as a proxy into the cluster for those Nodes.
+
+The given that the master service ip address is `10.0.0.1`, a Pod attempting to communicate with it's API server will do the following:
+
+- `GET https://10.0.0.1/` with it's private and public keys.
+- due to the proxy's sidecar, all traffic does to the `clisel-client` listening on `10.0.0.1`.
+- `chisel-client` has already opened a TCP connection to the upstream server and it creates a new SSH channel, sending the data to `chisel-server`.
+- `chisel-server` is assigned to a specific `kube-apiserver` and sends data to it.
+
+One problem with chisel is the lack of custom certificate authority support in the client, but there is an opened PR to [fix it](https://github.com/jpillora/chisel/pull/129).
```

Comment:
What is the alternative here if we decide to go with Chisel?

Reply:
What do you mean by alternative?

Reply:
I meant if there is an alternative for custom CA support, I can also have a look at the Chisel PR to understand more.

Reply:
Either `chisel-server` should serve a valid (from client point of view) certificate or no TLS should be served at all (which is not good from a security point of view)